importlib_resources/_common.py

```diff
@@ -87,13 +87,19 @@ def _(cand: None) -> types.ModuleType:
     return resolve(_infer_caller().f_globals['__name__'])
 
 
+@functools.lru_cache
+def _this_filename():
+    frame = inspect.currentframe()
+    return __file__ if frame is None else inspect.getframeinfo(frame).filename
```

Comment:
I'm slightly worried that the `inspect.currentframe() is None` case is not covered by tests. Also, I'm a little uneasy that this fallback still leaves other Python implementations subject to the reported issue (since it falls back to `__file__`).

According to [the docs](https://docs.python.org/3/library/inspect.html#inspect.currentframe), only `currentframe` is subject to this limitation, but it's surprising to me that a system would support `currentframe` but not `stack`, when stack would include the current frame.

But that gives me an idea - maybe the current filename could be extracted from the top of the stack, saving the need for an extra inspect call and conditional logic.

Reply:
I've done this in 5c3e38569, avoiding the conditional logic. This change now feels a lot more like a bugfix.